File: lang/csharp/cxlib/Table.cs
// CX C# binding — Public Table API per ADR 0018 D1.
//
// 17-member canonical Table API against the V core's :table blocks via
// the C ABI. Per ADR 0018 §D2: C# uses PascalCase methods matching the
// canonical surface.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Text.Json;

namespace CX;

/// <summary>Column iterator view (returned by <see cref="Table.IterCols"/>).</summary>
public sealed class ColumnView
{
    public string Name { get; }
    public string TypeName { get; }
    public IReadOnlyList<object?> Values { get; }

    public ColumnView(string name, string typeName, IReadOnlyList<object?> values)
    {
        Name = name;
        TypeName = typeName;
        Values = values;
    }
}

/// <summary>
/// Immutable handle over a single <c>:table</c> block. Implements the
/// 17-member canonical Table API (ADR 0018 §D1) with .NET PascalCase
/// naming. Construct via <see cref="FromCx"/>, <see cref="FromCxAll"/>,
/// or <see cref="Create"/>.
/// </summary>
public sealed class Table : IEnumerable<IReadOnlyDictionary<string, object?>>
{
    private readonly string[] _cols;
    private readonly string[] _types;
    private readonly object?[][] _rows;

    private Table(string[] cols, string[] types, object?[][] rows)
    {
        _cols = cols;
        _types = types;
        _rows = rows;
    }

    // ── Construction ─────────────────────────────────────────────────────────

    /// <summary>Parse CX source and return the first <c>:table</c> block.</summary>
    public static Table FromCx(string src)
    {
        var tables = FromCxAll(src);
        if (tables.Count == 0)
            throw new InvalidOperationException("cxlib: no :table block found in source");
        return tables[0];
    }

    /// <summary>Return every <c>:table</c> block in the source (preorder).</summary>
    public static IReadOnlyList<Table> FromCxAll(string src)
    {
        var payload = CxLib.ToDataBin(src);
        var decoded = DataBin.Decode(payload);
        var out_ = new List<Table>();
        CollectTables(decoded, out_);
        return out_;
    }

    /// <summary>Direct construction with 4-invariant validation per ADR 0018 §D7.</summary>
    public static Table Create(
        IReadOnlyList<string> cols,
        IReadOnlyList<string> types,
        IReadOnlyList<IReadOnlyList<object?>> rows)
    {
        if (cols.Count != types.Count)
            throw new ArgumentException(
                $"cxlib: len(cols)={cols.Count} != len(types)={types.Count}");
        var seen = new HashSet<string>();
        foreach (var c in cols)
        {
            if (!seen.Add(c))
                throw new ArgumentException($"cxlib: duplicate column name \"{c}\"");
        }
        var rowArr = new object?[rows.Count][];
        for (int i = 0; i < rows.Count; i++)
        {
            if (rows[i].Count != cols.Count)
                throw new ArgumentException(
                    $"cxlib: row {i} has {rows[i].Count} cells; expected {cols.Count}");
            rowArr[i] = rows[i].ToArray();
        }
        return new Table(cols.ToArray(), types.ToArray(), rowArr);
    }

    // ── Properties (4) ───────────────────────────────────────────────────────

    public IReadOnlyList<string> Cols => _cols;
    public IReadOnlyList<string> Types => _types;
    public int RowCount => _rows.Length;
    public int ColCount => _cols.Length;

    // ── Access (9) ───────────────────────────────────────────────────────────

    public IReadOnlyDictionary<string, object?> Row(int i)
    {
        if (i < 0 || i >= _rows.Length)
            throw new ArgumentOutOfRangeException(nameof(i),
                $"cxlib: row index {i} out of bounds [0, {_rows.Length})");
        var d = new Dictionary<string, object?>(_cols.Length);
        for (int c = 0; c < _cols.Length; c++) d[_cols[c]] = _rows[i][c];
        return d;
    }

    public IReadOnlyList<object?> Column(string name)
    {
        int idx = Array.IndexOf(_cols, name);
        if (idx < 0) throw new ArgumentException($"cxlib: unknown column \"{name}\"");
        return ColAt(idx);
    }

    public IReadOnlyList<object?> ColAt(int i)
    {
        if (i < 0 || i >= _cols.Length)
            throw new ArgumentOutOfRangeException(nameof(i),
                $"cxlib: column index {i} out of bounds [0, {_cols.Length})");
        var col = new object?[_rows.Length];
        for (int r = 0; r < _rows.Length; r++) col[r] = _rows[r][i];
        return col;
    }

    public object? Cell(int r, int c)
    {
        if (r < 0 || r >= _rows.Length)
            throw new ArgumentOutOfRangeException(nameof(r),
                $"cxlib: row index {r} out of bounds [0, {_rows.Length})");
        if (c < 0 || c >= _cols.Length)
            throw new ArgumentOutOfRangeException(nameof(c),
                $"cxlib: column index {c} out of bounds [0, {_cols.Length})");
        return _rows[r][c];
    }

    public object? CellByName(int r, string name)
    {
        int idx = Array.IndexOf(_cols, name);
        if (idx < 0) throw new ArgumentException($"cxlib: unknown column \"{name}\"");
        return Cell(r, idx);
    }

    public Table Slice(int start, int end)
    {
        if (start < 0 || start > _rows.Length)
            throw new ArgumentOutOfRangeException(nameof(start),
                $"cxlib: slice start {start} out of bounds");
        if (end < start || end > _rows.Length)
            throw new ArgumentOutOfRangeException(nameof(end),
                $"cxlib: slice end {end} out of bounds (start={start})");
        var rows = new object?[end - start][];
        for (int i = 0; i < rows.Length; i++) rows[i] = (object?[])_rows[start + i].Clone();
        return new Table(_cols, _types, rows);
    }

    public Table Head(int n) => Slice(0, Math.Min(Math.Max(n, 0), _rows.Length));

    public Table Tail(int n)
    {
        int start = Math.Max(0, _rows.Length - n);
        return Slice(start, _rows.Length);
    }

    /// <summary>
    /// <c>SelectCols</c> — renamed from canonical <c>Select</c> to mirror
    /// the other bindings' rename pattern (LINQ already owns <c>Select</c>
    /// in .NET, so the rename is also defensive).
    /// </summary>
    public Table SelectCols(IReadOnlyList<string> names)
    {
        var indices = new int[names.Count];
        var newCols = new string[names.Count];
        var newTypes = new string[names.Count];
        for (int i = 0; i < names.Count; i++)
        {
            int idx = Array.IndexOf(_cols, names[i]);
            if (idx < 0) throw new ArgumentException($"cxlib: unknown column \"{names[i]}\"");
            indices[i] = idx;
            newCols[i] = _cols[idx];
            newTypes[i] = _types[idx];
        }
        var newRows = new object?[_rows.Length][];
        for (int r = 0; r < _rows.Length; r++)
        {
            var row = new object?[indices.Length];
            for (int j = 0; j < indices.Length; j++) row[j] = _rows[r][indices[j]];
            newRows[r] = row;
        }
        return new Table(newCols, newTypes, newRows);
    }

    // ── Iteration (2) ────────────────────────────────────────────────────────

    public IEnumerator<IReadOnlyDictionary<string, object?>> GetEnumerator()
    {
        for (int i = 0; i < _rows.Length; i++) yield return Row(i);
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    public IEnumerable<ColumnView> IterCols()
    {
        for (int i = 0; i < _cols.Length; i++)
            yield return new ColumnView(_cols[i], _types[i], ColAt(i));
    }

    // ── Conversion (5) ───────────────────────────────────────────────────────

    public string ToCx()
    {
        var sb = new StringBuilder();
        sb.Append("[_ :table[");
        for (int i = 0; i < _cols.Length; i++)
        {
            if (i > 0) sb.Append(' ');
            sb.Append(_cols[i]);
            if (!string.IsNullOrEmpty(_types[i])) { sb.Append(':'); sb.Append(_types[i]); }
        }
        sb.Append("]\n");
        foreach (var row in _rows)
        {
            sb.Append("  ");
            for (int i = 0; i < row.Length; i++)
            {
                if (i > 0) sb.Append(' ');
                sb.Append(FormatCxCell(row[i]));
            }
            sb.Append('\n');
        }
        sb.Append("]\n");
        return sb.ToString();
    }

    public string ToCsv(char delim = ',')
    {
        var sb = new StringBuilder();
        for (int i = 0; i < _cols.Length; i++)
        {
            if (i > 0) sb.Append(delim);
            sb.Append(_cols[i]);
        }
        sb.Append("\r\n");
        foreach (var row in _rows)
        {
            for (int i = 0; i < row.Length; i++)
            {
                if (i > 0) sb.Append(delim);
                sb.Append(FormatCsvCell(row[i], delim));
            }
            sb.Append("\r\n");
        }
        return sb.ToString();
    }

    public string ToJson()
    {
        return JsonSerializer.Serialize(ToDictList());
    }

    public byte[] ToDataBin() => CxLib.ToDataBin(ToCx());

    public IReadOnlyList<IReadOnlyDictionary<string, object?>> ToDictList()
    {
        var list = new List<IReadOnlyDictionary<string, object?>>(_rows.Length);
        for (int i = 0; i < _rows.Length; i++) list.Add(Row(i));
        return list;
    }

    // ── Equality ─────────────────────────────────────────────────────────────

    public override bool Equals(object? obj)
    {
        if (obj is not Table other) return false;
        if (_cols.Length != other._cols.Length) return false;
        if (_rows.Length != other._rows.Length) return false;
        for (int i = 0; i < _cols.Length; i++)
        {
            if (_cols[i] != other._cols[i]) return false;
            if (_types[i] != other._types[i]) return false;
        }
        for (int r = 0; r < _rows.Length; r++)
        {
            if (_rows[r].Length != other._rows[r].Length) return false;
            for (int c = 0; c < _rows[r].Length; c++)
            {
                if (!CellEquals(_rows[r][c], other._rows[r][c])) return false;
            }
        }
        return true;
    }

    public override int GetHashCode()
    {
        var h = new HashCode();
        foreach (var c in _cols) h.Add(c);
        foreach (var t in _types) h.Add(t);
        h.Add(_rows.Length);
        return h.ToHashCode();
    }

    // ── Internal: walk decoded data_bin value to find tables ─────────────────

    private static void CollectTables(object? value, List<Table> outList)
    {
        if (value is null) return;
        if (value is IDictionary<string, object?> dict)
        {
            foreach (var child in dict.Values) CollectTables(child, outList);
            return;
        }
        if (value is IList list)
        {
            if (LooksLikeTable(list))
            {
                var first = (IDictionary<string, object?>)list[0]!;
                var cols = first.Keys.OrderBy(k => k, StringComparer.Ordinal).ToArray();
                var types = cols.Select(_ => "").ToArray();
                var rows = new List<IReadOnlyList<object?>>(list.Count);
                foreach (var item in list)
                {
                    var d = (IDictionary<string, object?>)item!;
                    var row = new object?[cols.Length];
                    for (int c = 0; c < cols.Length; c++)
                        row[c] = d.TryGetValue(cols[c], out var v) ? v : null;
                    rows.Add(row);
                }
                outList.Add(Create(cols, types, rows));
                return;
            }
            foreach (var child in list) CollectTables(child, outList);
        }
    }

    private static bool LooksLikeTable(IList list)
    {
        if (list.Count == 0) return false;
        if (list[0] is not IDictionary<string, object?> first) return false;
        var keys = new HashSet<string>(first.Keys);
        if (keys.Count == 0) return false;
        foreach (var item in list)
        {
            if (item is not IDictionary<string, object?> d) return false;
            if (d.Count != keys.Count) return false;
            foreach (var k in d.Keys) if (!keys.Contains(k)) return false;
        }
        return true;
    }

    // ── Internal: cell formatters / equality ─────────────────────────────────

    private static string FormatCxCell(object? v)
    {
        if (v is null) return "null";
        if (v is bool b) return b ? "true" : "false";
        if (v is long l) return l.ToString(CultureInfo.InvariantCulture);
        if (v is int i) return i.ToString(CultureInfo.InvariantCulture);
        if (v is double d) return d.ToString("R", CultureInfo.InvariantCulture);
        if (v is float f) return f.ToString("R", CultureInfo.InvariantCulture);
        if (v is IList<object?> list)
        {
            var sb = new StringBuilder("[");
            for (int k = 0; k < list.Count; k++)
            {
                if (k > 0) sb.Append(", ");
                sb.Append(FormatCxCell(list[k]));
            }
            sb.Append(']');
            return sb.ToString();
        }
        if (v is IDictionary<string, object?> dict)
        {
            var keys = dict.Keys.OrderBy(k => k, StringComparer.Ordinal).ToArray();
            var sb = new StringBuilder("{");
            for (int k = 0; k < keys.Length; k++)
            {
                if (k > 0) sb.Append(", ");
                sb.Append(FormatCxKey(keys[k]));
                sb.Append(": ");
                sb.Append(FormatCxCell(dict[keys[k]]));
            }
            sb.Append('}');
            return sb.ToString();
        }
        string s = Convert.ToString(v, CultureInfo.InvariantCulture) ?? "";
        if (s.Length == 0 || NeedsQuote(s))
            return "'" + s.Replace("'", "''") + "'";
        return s;
    }

    private static bool NeedsQuote(string s)
    {
        foreach (var ch in s)
        {
            if (char.IsWhiteSpace(ch)) return true;
            if (ch == '\'' || ch == '[' || ch == ']' || ch == '(' || ch == ')'
                || ch == '{' || ch == '}' || ch == ',') return true;
        }
        return false;
    }

    private static string FormatCxKey(string k)
    {
        bool simple = k.Length > 0;
        for (int i = 0; i < k.Length && simple; i++)
        {
            char ch = k[i];
            bool ok = (ch == '_')
                || (ch >= 'a' && ch <= 'z')
                || (ch >= 'A' && ch <= 'Z')
                || (i > 0 && ch >= '0' && ch <= '9');
            if (!ok) simple = false;
        }
        if (simple) return k;
        return "'" + k.Replace("'", "''") + "'";
    }

    private static string FormatCsvCell(object? v, char delim)
    {
        if (v is null) return "";
        if (v is IDictionary<string, object?> || v is IList<object?>)
        {
            string json = JsonSerializer.Serialize(v);
            return "\"" + json.Replace("\"", "\"\"") + "\"";
        }
        if (v is bool b) return b ? "true" : "false";
        string s = Convert.ToString(v, CultureInfo.InvariantCulture) ?? "";
        if (s.IndexOf(delim) >= 0 || s.IndexOf('"') >= 0 ||
            s.IndexOf('\n') >= 0 || s.IndexOf('\r') >= 0)
        {
            return "\"" + s.Replace("\"", "\"\"") + "\"";
        }
        return s;
    }

    private static bool CellEquals(object? a, object? b)
    {
        if (ReferenceEquals(a, b)) return true;
        if (a is null || b is null) return false;
        if (a is IList<object?> la && b is IList<object?> lb)
        {
            if (la.Count != lb.Count) return false;
            for (int i = 0; i < la.Count; i++)
                if (!CellEquals(la[i], lb[i])) return false;
            return true;
        }
        if (a is IDictionary<string, object?> da && b is IDictionary<string, object?> db)
        {
            if (da.Count != db.Count) return false;
            foreach (var kv in da)
            {
                if (!db.TryGetValue(kv.Key, out var bv)) return false;
                if (!CellEquals(kv.Value, bv)) return false;
            }
            return true;
        }
        return a.Equals(b);
    }
}

File: lang/typescript/cxlib/src/table.ts
/**
 * CX TypeScript binding — Public Table API per ADR 0018 D1.
 *
 * Implements the 17-member canonical Table API against the V core's
 * :table blocks via the C ABI. Per ADR 0018 §D2: TS uses camelCase
 * methods matching the canonical surface.
 */

import { toDataBin as _toDataBin } from "./index";
import * as data_bin from "./data_bin";

type CellValue =
  | null
  | boolean
  | number
  | string
  | CellValue[]
  | { [key: string]: CellValue };

/** Immutable handle over a single :table block. */
export class Table {
  private readonly _cols: string[];
  private readonly _types: string[];
  private readonly _rows: CellValue[][];

  private constructor(
    cols: string[],
    types: string[],
    rows: CellValue[][],
  ) {
    this._cols = [...cols];
    this._types = [...types];
    this._rows = rows.map((r) => [...r]);
  }

  // ── Construction ─────────────────────────────────────────────────────────

  /** Parse CX source and return the first :table block. */
  static fromCx(src: string): Table {
    const tables = Table.fromCxAll(src);
    if (tables.length === 0) {
      throw new Error("cxlib: no :table block found in source");
    }
    return tables[0];
  }

  /** Return every :table block in the source. */
  static fromCxAll(src: string): Table[] {
    const payload = _toDataBin(src);
    const decoded = data_bin.decode(payload);
    const out: Table[] = [];
    collectTables(decoded, out);
    return out;
  }

  /** Direct construction with 4-invariant validation per ADR 0018 §D7. */
  static create(
    cols: string[],
    types: string[],
    rows: CellValue[][],
  ): Table {
    if (cols.length !== types.length) {
      throw new Error(
        `cxlib: len(cols)=${cols.length} != len(types)=${types.length}`,
      );
    }
    const seen = new Set<string>();
    for (const c of cols) {
      if (seen.has(c)) {
        throw new Error(`cxlib: duplicate column name "${c}"`);
      }
      seen.add(c);
    }
    for (let i = 0; i < rows.length; i++) {
      if (rows[i].length !== cols.length) {
        throw new Error(
          `cxlib: row ${i} has ${rows[i].length} cells; expected ${cols.length}`,
        );
      }
    }
    return new Table(cols, types, rows);
  }

  // ── Properties (4) — exposed as getters ──────────────────────────────────

  get cols(): string[] { return [...this._cols]; }
  get types(): string[] { return [...this._types]; }
  get rowCount(): number { return this._rows.length; }
  get colCount(): number { return this._cols.length; }

  // ── Access (9) ───────────────────────────────────────────────────────────

  row(i: number): { [key: string]: CellValue } {
    if (i < 0 || i >= this._rows.length) {
      throw new Error(
        `cxlib: row index ${i} out of bounds [0, ${this._rows.length})`,
      );
    }
    const out: { [key: string]: CellValue } = {};
    this._cols.forEach((name, c) => {
      out[name] = this._rows[i][c];
    });
    return out;
  }

  column(name: string): CellValue[] {
    const idx = this._cols.indexOf(name);
    if (idx < 0) throw new Error(`cxlib: unknown column "${name}"`);
    return this.colAt(idx);
  }

  colAt(i: number): CellValue[] {
    if (i < 0 || i >= this._cols.length) {
      throw new Error(
        `cxlib: column index ${i} out of bounds [0, ${this._cols.length})`,
      );
    }
    return this._rows.map((row) => row[i]);
  }

  cell(r: number, c: number): CellValue {
    if (r < 0 || r >= this._rows.length) {
      throw new Error(
        `cxlib: row index ${r} out of bounds [0, ${this._rows.length})`,
      );
    }
    if (c < 0 || c >= this._cols.length) {
      throw new Error(
        `cxlib: column index ${c} out of bounds [0, ${this._cols.length})`,
      );
    }
    return this._rows[r][c];
  }

  cellByName(r: number, name: string): CellValue {
    const idx = this._cols.indexOf(name);
    if (idx < 0) throw new Error(`cxlib: unknown column "${name}"`);
    return this.cell(r, idx);
  }

  slice(start: number, end: number): Table {
    if (start < 0 || start > this._rows.length) {
      throw new Error(`cxlib: slice start ${start} out of bounds`);
    }
    if (end < start || end > this._rows.length) {
      throw new Error(`cxlib: slice end ${end} out of bounds (start=${start})`);
    }
    return new Table(this._cols, this._types, this._rows.slice(start, end));
  }

  head(n: number): Table {
    return this.slice(0, Math.min(Math.max(n, 0), this._rows.length));
  }

  tail(n: number): Table {
    const start = Math.max(0, this._rows.length - n);
    return this.slice(start, this._rows.length);
  }

  /**
   * `selectCols` — renamed from canonical `select` to mirror the
   * other bindings' rename pattern (Rust/V/Python use select_cols).
   * TS doesn't have a stdlib `select` conflict, but consistency wins.
   */
  selectCols(names: string[]): Table {
    const indices: number[] = [];
    const newCols: string[] = [];
    const newTypes: string[] = [];
    for (const name of names) {
      const idx = this._cols.indexOf(name);
      if (idx < 0) throw new Error(`cxlib: unknown column "${name}"`);
      indices.push(idx);
      newCols.push(this._cols[idx]);
      newTypes.push(this._types[idx]);
    }
    const newRows = this._rows.map((row) => indices.map((i) => row[i]));
    return new Table(newCols, newTypes, newRows);
  }

  // ── Iteration (2) ────────────────────────────────────────────────────────

  /** `for (const row of t)` yields ordered objects in row order. */
  *[Symbol.iterator](): IterableIterator<{ [key: string]: CellValue }> {
    for (let i = 0; i < this._rows.length; i++) {
      yield this.row(i);
    }
  }

  *iterCols(): IterableIterator<ColumnView> {
    for (let i = 0; i < this._cols.length; i++) {
      yield {
        name: this._cols[i],
        typeName: this._types[i],
        values: this.colAt(i),
      };
    }
  }

  // ── Conversion (5) ───────────────────────────────────────────────────────

  toCx(): string {
    const header = this._cols
      .map((name, i) =>
        this._types[i] === "" ? name : `${name}:${this._types[i]}`
      )
      .join(" ");
    const lines = [`[_ :table[${header}]`];
    for (const row of this._rows) {
      const cells = row.map(formatCxCell).join(" ");
      lines.push(`  ${cells}`);
    }
    lines.push("]");
    return lines.join("\n") + "\n";
  }

  toCsv(delim: string = ","): string {
    if (delim.length !== 1) {
      throw new Error(`cxlib: toCsv delim must be 1 char; got ${delim.length}`);
    }
    const lines = [this._cols.join(delim)];
    for (const row of this._rows) {
      lines.push(row.map((v) => formatCsvCell(v, delim)).join(delim));
    }
    return lines.join("\r\n") + "\r\n";
  }

  toJson(): string {
    return JSON.stringify(this.toDictList());
  }

  toDataBin(): Buffer {
    return _toDataBin(this.toCx());
  }

  toDictList(): { [key: string]: CellValue }[] {
    return this._rows.map((_, i) => this.row(i));
  }

  // ── Equality ─────────────────────────────────────────────────────────────

  equals(other: Table): boolean {
    if (!other || !(other instanceof Table)) return false;
    if (this._cols.length !== other._cols.length) return false;
    if (this._rows.length !== other._rows.length) return false;
    for (let i = 0; i < this._cols.length; i++) {
      if (this._cols[i] !== other._cols[i]) return false;
      if (this._types[i] !== other._types[i]) return false;
    }
    return JSON.stringify(this._rows) === JSON.stringify(other._rows);
  }
}

/** Column iterator view. */
export interface ColumnView {
  name: string;
  typeName: string;
  values: CellValue[];
}

// ── Internal: walk a decoded data_bin value to find tables ───────────────────

function collectTables(value: any, out: Table[]): void {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    for (const child of Object.values(value)) {
      collectTables(child, out);
    }
  } else if (Array.isArray(value)) {
    if (looksLikeTable(value)) {
      const first = value[0] as { [key: string]: any };
      const cols = Object.keys(first).sort();
      const types = cols.map(() => "");
      const rows = value.map((item) =>
        cols.map((c) => (item as any)[c])
      );
      out.push((Table as any).create(cols, types, rows));
    } else {
      for (const child of value) {
        collectTables(child, out);
      }
    }
  }
}

function looksLikeTable(value: any[]): boolean {
  if (value.length === 0) return false;
  const first = value[0];
  if (!first || typeof first !== "object" || Array.isArray(first)) {
    return false;
  }
  const keys = new Set(Object.keys(first));
  if (keys.size === 0) return false;
  return value.every((item) => {
    if (!item || typeof item !== "object" || Array.isArray(item)) return false;
    const itemKeys = Object.keys(item);
    if (itemKeys.length !== keys.size) return false;
    return itemKeys.every((k) => keys.has(k));
  });
}

// ── Internal: cell formatters ────────────────────────────────────────────────

function formatCxCell(v: CellValue): string {
  if (v === null) return "null";
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "number") return String(v);
  if (Array.isArray(v)) {
    return "[" + v.map(formatCxCell).join(", ") + "]";
  }
  if (typeof v === "object") {
    const keys = Object.keys(v).sort();
    return (
      "{" +
      keys.map((k) => `${formatCxKey(k)}: ${formatCxCell(v[k])}`).join(", ") +
      "}"
    );
  }
  // String
  const s = String(v);
  if (s === "" || /[\s'\[\](){},]/.test(s)) {
    return "'" + s.replace(/'/g, "''") + "'";
  }
  return s;
}

function formatCxKey(k: string): string {
  if (/^[a-zA-Z_][a-zA-Z0-9_]*$/.test(k)) return k;
  return "'" + k.replace(/'/g, "''") + "'";
}

function formatCsvCell(v: CellValue, delim: string): string {
  if (v === null) return "";
  if (typeof v === "object") {
    const json = JSON.stringify(v);
    return '"' + json.replace(/"/g, '""') + '"';
  }
  if (typeof v === "boolean") return v ? "true" : "false";
  const s = String(v);
  if (s.includes(delim) || s.includes('"') || s.includes("\n") || s.includes("\r")) {
    return '"' + s.replace(/"/g, '""') + '"';
  }
  return s;
}
